### PR TITLE
Client server branch

### PR DESCRIPTION
**Pull request by moyaog**
Contains the client and server classes. Variables like KEYSTORE, PASSWORD, CLIENT_KEY_LOC, HOST, and PORT are hard coded values that are in a class called HardCodes. Instructions will be added to the README on how to create your own HardCodes class before running the code.

### CODE REVIEW

File: InitSSL.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
+import static my.HardCodes.*;
+
+import java.io.*;
+import java.security.KeyStore;
+import javax.net.ssl.KeyManagerFactory;
+import javax.net.ssl.SSLContext;
+import javax.net.ssl.TrustManagerFactory;
+
+public class InitSSL {
```
Comment (a-palchikov) [inline InitSSL.java:9]: I'd try to avoid naming duplications: `InitSSL.InitSsl()` is not meaningful and is redundant.
You could've named the class `Credentials` and the method `init` or `create`.

File: InitSSL.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+      TrustManagerFactory trustManagerFactory = TrustManagerFactory.getInstance(TRUST_MANAGER);
+      trustManagerFactory.init(keyStore);
+
+      SSLContext sslContext = SSLContext.getInstance(PROTOCOL_VER);
+      sslContext.init(keyManagerFactory.getKeyManagers(), trustManagerFactory.getTrustManagers(), null);
+
+      return sslContext;
+    } catch(Exception e) {
+      e.printStackTrace();
```
Comment (a-palchikov) [inline InitSSL.java:26]: This is not too helpful for the caller as it is not possible to derive reasons for failure given just a plain `null` result. Can you elaborate what might fail in this method and leave a comment about it?

File: Client.java [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+import javax.net.ssl.SSLSession;
+import javax.net.ssl.SSLSocket;
+import javax.net.ssl.SSLSocketFactory;
+
+public class Client {
+  private final static int COMMAND_LOC = 0;
+  private final static int PID_OR_PROCESS_LOC = 1;
+  private final static int MIN_ARGS = 2;
+
+  public static void main(String[] args) throws JSONException {
```
Comment (a-palchikov) [inline Client.java]: There's hardly any structure in `main` - it is extremely hard to follow (and will also be hard for a different developer to modify). Have you considered breaking it into smaller pieces and structuring `main` to outline the high-level flow of the client?
For me, it could be as follows:
```java
public class Client {
  private final static int COMMAND_LOC = 0;
  private final static int PID_OR_PROCESS_LOC = 1;
  private final static int MIN_ARGS = 2;

  public static void main(String[] args) throws JSONException {
    ServerResponse resp;
    try {
      printInstructionsBanner();
      JSONObject request = readRequestFromStdin();
      resp = submitServerRequest(request);
      if(!resp.isValid(requestID(request))) {
        throw new Exception("Response message received did not match ID of request message sent");
      }
      handleResponse(resp);
    } catch (Exception e) {
      e.printStackTrace();
    } finally {
      if(resp != null) {
        resp.close();
      }
    }
  }

  private void printInstructionsBanner() {
    System.out.println("You may start a job, stop a job by PID, query a job by PID," + 
      " get the output of a running job process, or get a list of all current running jobs");
    System.out.println("Valid commands are: START <process>, STOP <pid>, QUERY <pid>," +
      " OUTPUT <pid>, and CURRENT");
  }

  private JSONObject readRequestFromStdin() throws Exception {
    BuildJson buildJson = new BuildJson();
    Request request = new Request();

    Scanner scan = new Scanner(System.in);
    String[] split = System.console().readLine().split(" ");
    System.out.println();

    if(split[COMMAND_LOC].toLowerCase().equals(START)) {
      if(split.length < MIN_ARGS) {
        throw new Exception("No process provided");
      }
      String completeJob = split[PID_OR_PROCESS_LOC];
      for(int i = PID_OR_PROCESS_LOC+1; i < split.length; i++) {
        completeJob += (" " + split[i]);
      }
      request.setProcess(completeJob);
      return buildJson.buildRequest(request, START);
    } else if(split[COMMAND_LOC].toLowerCase().equals(STOP)) {
      if(split.length < MIN_ARGS) {
        throw new Exception("No PID provided");
      }
      long pid = Long.parseLong(split[PID_OR_PROCESS_LOC]);
      request.setPid(pid);
      return buildJson.buildRequest(request, STOP);
    } else if(split[COMMAND_LOC].toLowerCase().equals(QUERY)) {
      if(split.length < MIN_ARGS) {
        throw new Exception("No PID provided");
      }
      long pid = Long.parseLong(split[PID_OR_PROCESS_LOC]);
      request.setPid(pid);
      return buildJson.buildRequest(request, QUERY);
    } else if(split[COMMAND_LOC].toLowerCase().equals(GET_CURRENT)) {
      return buildJson.buildRequest(request, GET_CURRENT);
    } else if(split[COMMAND_LOC].toLowerCase().equals(OUTPUT)) {
      if(split.length < MIN_ARGS) {
        throw new Exception("No PID provided");
      }
      long pid = Long.parseLong(split[PID_OR_PROCESS_LOC]);
      request.setPid(pid);
      return buildJson.buildRequest(request, OUTPUT);
    } else {
      throw new Exception("No valid command provided");
    }
  }

  private ServerResponse submitServerRequest(JSONObject request) {
    Credentials creds = new Credentials();
    SSLContext sslContext = creds.init(CLIENT_KEY_LOC);
    SSLSocketFactory sslSocketFactory = sslContext.getSocketFactory();
    SSLSocket sslSocket = (SSLSocket)sslSocketFactory.createSocket(HOST, PORT);
    sslSocket.setEnabledCipherSuites(sslSocket.getSupportedCipherSuites());
    sslSocket.startHandshake();
    SSLSession sslSession = sslSocket.getSession();

    OutputStream outputStream = sslSocket.getOutputStream();
    ObjectOutputStream objectOutputStream = new ObjectOutputStream(outputStream);
    objectOutputStream.writeObject(request.toString());
    outputStream.flush();

    InputStream inputStream = sslSocket.getInputStream();
    ObjectInputStream objectInputStream = new ObjectInputStream(inputStream);

    JSONObject jsonObjParsed = new JSONObject((String)objectInputStream.readObject());
    ParseJson parseJson = new ParseJson();
    HashMap<String, Object> resp = parseJson.parseJson(jsonObjParsed);
    return new ServerResponse(resp, sslSocket);
  }

  public void handleResponse(ServerResponse resp) throws Exception {
    try{
      ErrorInfo errorInfo = resp.getErrorInfo();
      Array<ErrorInfo> runningJobs = resp.getRunningJobs();
      if(errorInfo == null && runningJobs == null) {
        throw new Exception("Response is empty");
      }
      if(errorInfo != null) {
        if(errorInfo.getErrorCode() == ERR_SUCCESS) {
          System.out.println(errorInfo.getIoMessage());
          System.out.println(errorInfo.getErrorMessage());
          if(errorInfo.getPid() != null) {
            System.out.println("PID " + errorInfo.getPid());
          }
        } else {
          if(errorInfo.getErrorCode() == ERR_FAILED_TO_KILL) {
            System.out.println("Error: Failed to kill process");
          } else if(errorInfo.getErrorCode() == ERR_FAILED_TO_FIND) {
            System.out.println("Error: Failed to find process");
          } else if(errorInfo.getErrorCode() == ERR_FAILED_TO_START) {
            System.out.println("Error: Failed to start process");
          } else {
            System.out.println("Error: Unidentified error. Check error message for more information.");
          }
          System.out.println("Error Code: " + errorInfo.getErrorCode());
          System.out.println("Error Message: " + errorInfo.getErrorMessage());
        }
        return;
      }

      if(errorInfoArray.size() > 0) {
        System.out.println("Currently running processes:");
        for(int i = 0; i < errorInfoArray.size(); i++) {
          ErrorInfo tempErrorInfo = errorInfoArray.get(i);
          System.out.println("PID: " + tempErrorInfo.getPid() + "\tMessage: " + tempErrorInfo.getIoMessage());
        }
      } else {
        System.out.println("There are no currently running processes");
      }
    } catch(Exception e) {
      System.out.println(e.getClass().getName());
      System.out.println(e.getMessage());
    }
  }

  private String requestID(JSONObject request) {
    return (String)request.get(ID);
  }
}

class ServerResponse {
  HashMap<String, Object> response;
  SSLSocket socket;

  ServerResponse(HashMap<String, Object> resp, SSLSocket socket) {
    this.response = resp;
    this.socket = socket;
  }

  bool isValid(String id) {
    String responseId = (String)this.response.get(ID);
    return id.equals(responseId);
  }

  ErrorInfo getErrorInfo() {
    Response response = (Response)this.response.get(RESULT);
    return response.getErrorInfo();
  }

  ArrayList<ErrorInfo> getRunningJobs() {
    Response response = (Response)this.response.get(RESULT);
    return response.getRunningJobs();
  }

  void close() throws Exception {
    this.socket.close();
  }
}
```
It is still not ideal but you should get the point.
How would you test your code, btw?

It definitely applies to `Server` as well.

File: Server.java [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+import javax.net.ssl.SSLServerSocketFactory;
+import javax.net.ssl.SSLServerSocket;
+import javax.net.ssl.SSLSocket;
+
+public class Server {
+  public static void main(String[] args) throws JSONException, IOException {
+    ArrayList<ErrorInfo> pids = new ArrayList<ErrorInfo>();
+  
+    try {
+      InitSSL initSsl = new InitSSL();
```
Comment (russjones) [inline Server.java]: Ideally you would not be working at this low of a level and instead use something like HTTP/JSON or GRPC where a lot of this work would be abstracted away from you.

However, since it appears you are implementing a protocol (JSON-RPC), you should abstract this out yourself. 

Most of this code should sit in it's own class with a simple interface that you can use to perform a request and get a response. That interface is what you should use here.

File: InitSSL.java [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+import java.security.KeyStore;
+import javax.net.ssl.KeyManagerFactory;
+import javax.net.ssl.SSLContext;
+import javax.net.ssl.TrustManagerFactory;
+
+public class InitSSL {
+  SSLContext initSsl(String keyLocation) {
+    try {
+      KeyStore keyStore = KeyStore.getInstance(KEY_STORE);
+      keyStore.load(new FileInputStream(keyLocation), PASSWORD.toCharArray());
```
Comment (russjones) [inline InitSSL.java:13]: What is `PASSWORD` in this case?

Comment (moyaog) [reply InitSSL.java:13]: PASSWORD (and KEY_STORE) are hardcoded values in a class called HardCodes that I have not checked in. I also updated the PR description with more information.

File: InitSSL.java (hunk 1 above)
Comment (moyaog) [reply InitSSL.java:9]: Okay. I've renamed this class Credentials.

File: InitSSL.java (hunk 2 above)
Comment (moyaog) [reply InitSSL.java:26]: I've decided to remove the try-catch in this method, and I'm just opting to have the method throw an exception. If an exception is thrown in this method, I want the server to fail to run. With this implementation, the Server will catch the exception this method throws and print out information.

File: Client.java (hunk 3 above)
Comment (moyaog) [reply Client.java]: I've followed your suggestion and restructured Server and Client. I've also added classes called ParsedRequest and ParsedResponse.

File: Server.java (hunk 4 above)
Comment (moyaog) [reply Server.java]: I'm not sure if this still applies with the changes that I've made. Let me know.

File: Client.java [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+      parsedResponse = sendJsonRequest(jsonRequest);
+
+      if(!parsedResponse.isValid(requestId(jsonRequest))) {
+        throw new Exception("Response message received did not match ID of request message sent");
+      }
+
+      handleResponse(parsedResponse);
+
+      if(parsedResponse.isInitialized()) 
+        parsedResponse.close();
```
Comment (a-palchikov) [inline Client.java]: I believe this is still not exception-safe as it will leave the socket open with either an exception when validating the response ID or any exception from `handleResponse`

File: Client.java [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+      json = buildJson.buildRequest(request, GET_CURRENT);
+    } else if(split[COMMAND_LOC].toLowerCase().equals(OUTPUT)) {
+      if(split.length < MIN_ARGS) {
+        throw new Exception("No PID provided");
+      }
+      long pid = Long.parseLong(split[PID_OR_PROCESS_LOC]);
+      request.setPid(pid);
+      json = buildJson.buildRequest(request, OUTPUT);
+    } else {
+      throw new Exception("No valid command provided");
```
Comment (a-palchikov) [inline Client.java:89]: nit: this should be the last statement of the method - not sure why you'd want to store the result into a variable instead of returning the results as soon as they're available.

File: Client.java [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+    outputStream.flush();
+
+    InputStream inputStream = sslSocket.getInputStream();
+    ObjectInputStream objectInputStream = new ObjectInputStream(inputStream);      
+   
+    JSONObject jsonObjParsed = new JSONObject((String)objectInputStream.readObject());
+    ParseJson parseJson = new ParseJson();
+    HashMap<String, Object> responseMap = parseJson.parseJson(jsonObjParsed);
+
+    return new ParsedResponse(responseMap, sslSocket);
```
Comment (a-palchikov) [inline Client.java]: I just realized that socket will not be needed outside this method - do you think you need to have it around and have to explicitly close it only after handling the response?
This equally applies to `ParsedRequest`.
Sorry if my previous comment was confusing - I sketched the code in a hurry.

**Review by a-palchikov — APPROVED** [review]
This will be good to go if you address the last couple of comments about the socket.

File: Client.java (hunk 7 above)
Comment (moyaog) [reply Client.java:89]: Fixing.

File: Client.java (hunk 8 above)
Comment (moyaog) [reply Client.java]: I have fixed this issued in ParsedResponse. The socket is needed in multiple methods in the server so I have not made the equivalent change to ParsedRequest.

**Comment by moyaog** [comment]
I am making a few new helper methods in Client and Server in order to help with testing. They are now checked in.

File: Client.java (hunk 6 above)
Comment (moyaog) [reply Client.java]: I have tried to fix this issue in Client and Server with some changes that I have made. I will push that code shortly.
